Ehlers Damage test failing sporadically

Symptom: `ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension` sometimes falies (roughly 50% of the time)
```
[8] lehmannc@envinf141l:build-release-eigenlis$ OMP_NUM_THREADS=1 ctest -R ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
Test project /home/lehmannc/prog/ogs6/github-chleh-PRs/build-release-eigenlis
    Start 210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
1/2 Test #210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension ...........***Failed    9.26 sec
    Start 211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff
2/2 Test #211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff ...***Failed    0.05 sec

0% tests passed, 2 tests failed out of 2

Total Test time (real) =   9.33 sec

The following tests FAILED:
	210 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension (Failed)
	211 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff (Failed)
Errors while running CTest
[8] lehmannc@envinf141l:build-release-eigenlis$ OMP_NUM_THREADS=1 ctest -R ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
Test project /home/lehmannc/prog/ogs6/github-chleh-PRs/build-release-eigenlis
    Start 210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
1/2 Test #210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension ...........***Failed    9.42 sec
    Start 211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff
2/2 Test #211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff ...***Failed    0.05 sec

0% tests passed, 2 tests failed out of 2

Total Test time (real) =   9.50 sec

The following tests FAILED:
	210 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension (Failed)
	211 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff (Failed)
Errors while running CTest
[8] lehmannc@envinf141l:build-release-eigenlis$ OMP_NUM_THREADS=1 ctest -R ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
Test project /home/lehmannc/prog/ogs6/github-chleh-PRs/build-release-eigenlis
    Start 210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
1/2 Test #210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension ...........***Failed    0.38 sec
    Start 211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff
2/2 Test #211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff ...***Failed    0.03 sec

0% tests passed, 2 tests failed out of 2

Total Test time (real) =   0.43 sec

The following tests FAILED:
	210 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension (Failed)
	211 - ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff (Failed)
Errors while running CTest
[8] lehmannc@envinf141l:build-release-eigenlis$ OMP_NUM_THREADS=1 ctest -R ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
Test project /home/lehmannc/prog/ogs6/github-chleh-PRs/build-release-eigenlis
    Start 210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension
1/2 Test #210: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension ...........   Passed   10.72 sec
    Start 211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff
2/2 Test #211: ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension-vtkdiff ...   Passed    0.05 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =  10.79 sec
```

Probably this is linked to the following compiler warnings: (gcc 7.1.1, dynamic shape matrices):
```
[ 63%] Building CXX object MaterialLib/CMakeFiles/MaterialLib.dir/SolidModels/Ehlers.cpp.o
In file included from /home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.cpp:10:0:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.h: In instantiation of ‘std::unique_ptr<typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables> MaterialLib::Solids::Lubby2::Lubby2<DisplacementDim>::createMaterialStateVariables() [with int DisplacementDim = 2; typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables = MaterialLib::Solids::MechanicsBase<2>::MaterialStateVariables]’:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.cpp:19:16:   required from here
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.h:160:13: warning: ‘new’ of type ‘MaterialLib::Solids::Lubby2::Lubby2<2>::MaterialStateVariables’ with extended alignment 32 [-Waligned-new=]
             new MaterialStateVariables};
             ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.h:160:13: note: uses ‘void* operator new(std::size_t)’, which does not have an alignment parameter
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.h:160:13: note: use ‘-faligned-new’ to enable C++17 over-aligned new support
In file included from /home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.cpp:11:0:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2-impl.h: In instantiation of ‘boost::optional<std::tuple<Eigen::Matrix<double, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1, 0, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1>, std::unique_ptr<typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables, std::default_delete<typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables> >, Eigen::Matrix<double, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value> > > MaterialLib::Solids::Lubby2::Lubby2<DisplacementDim>::integrateStress(double, const ProcessLib::SpatialPosition&, double, const KelvinVector&, const KelvinVector&, const KelvinVector&, const typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables&) [with int DisplacementDim = 2; typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables = MaterialLib::Solids::MechanicsBase<2>::MaterialStateVariables; MaterialLib::Solids::Lubby2::Lubby2<DisplacementDim>::KelvinVector = Eigen::Matrix<double, 4, 1>]’:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2.cpp:19:16:   required from here
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2-impl.h:191:13: warning: ‘new’ of type ‘MaterialLib::Solids::Lubby2::Lubby2<2>::MaterialStateVariables’ with extended alignment 32 [-Waligned-new=]
             new MaterialStateVariables{state}},
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2-impl.h:191:13: note: uses ‘void* operator new(std::size_t)’, which does not have an alignment parameter
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Lubby2-impl.h:191:13: note: use ‘-faligned-new’ to enable C++17 over-aligned new support
In file included from /home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers.cpp:11:0:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers-impl.h: In instantiation of ‘boost::optional<std::tuple<Eigen::Matrix<double, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1, 0, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1>, std::unique_ptr<typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables, std::default_delete<typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables> >, Eigen::Matrix<double, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, 1, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value, ProcessLib::KelvinVectorDimensions<DisplacementDim>::value> > > MaterialLib::Solids::Ehlers::SolidEhlers<DisplacementDim>::integrateStress(double, const ProcessLib::SpatialPosition&, double, const KelvinVector&, const KelvinVector&, const KelvinVector&, const typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables&) [with int DisplacementDim = 2; typename MaterialLib::Solids::MechanicsBase<DisplacementDim>::MaterialStateVariables = MaterialLib::Solids::MechanicsBase<2>::MaterialStateVariables; MaterialLib::Solids::Ehlers::SolidEhlers<DisplacementDim>::KelvinVector = Eigen::Matrix<double, 4, 1>]’:
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers.cpp:19:16:   required from here
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers-impl.h:738:13: warning: ‘new’ of type ‘MaterialLib::Solids::Ehlers::StateVariables<2>’ with extended alignment 32 [-Waligned-new=]
             new StateVariables<DisplacementDim>{
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                 static_cast<StateVariables<DisplacementDim> const&>(state)}},
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers-impl.h:738:13: note: uses ‘void* operator new(std::size_t)’, which does not have an alignment parameter
/home/lehmannc/prog/ogs6/github-chleh-PRs/src/MaterialLib/SolidModels/Ehlers-impl.h:738:13: note: use ‘-faligned-new’ to enable C++17 over-aligned new support
```

Because adding `add_definitions(-DEIGEN_DONT_VECTORIZE)` to the main `CMakeLists.txt` fixes this problem: Now `$ OMP_NUM_THREADS=1 ctest --repeat-until-fail 50 -j 3 -R ogs-Mechanics_PlasticModel_SDED_Ehlers_Damage_UniaxialTension` just runs fine.

@endJunction noticed that in some classes `EIGEN_MAKE_ALIGNED_OPERATOR_NEW` is missing.

This problem adds to a set of seven issues or pull requests regarding Eigen alignment: #743, #1221, #1250, #1323, #1483, #1684, #1760. Maybe we should consider disabling Eigen alignment entirely (at least by default).
Additionally, the warnings above are generated only with very recent compiler versions (cf. https://gcc.gnu.org/gcc-7/changes.html). So the average user might not get notified of the indicated problem.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -90,6 +90,13 @@ option(OGS_USE_EIGEN_UNSUPPORTED "Use Eigen unsupported modules" ON)
 option(OGS_EIGEN_INITIALIZE_MATRICES_BY_NAN "" ON)
 option(EIGEN_NO_DEBUG "Disables Eigen's assertions" OFF)
 
+# We assume that it's save to use vectorization with Eigen < 3.3 (strictly smaller than 3.3.!).
+# At least we only observed vectorization issues with Eigen 3.3.x.
+# If you want to use Eigen vectorization, make sure that you run all the ctests several times, e.g.:
+# $ ctest --repeat-until-fail 50
+# You might also want to take a look at https://github.com/ufz/ogs/issues/1881.
+option(EIGEN_DONT_VECTORIZE "Disables explicit vectorization when defined." ON)
+
 set(OGS_EIGEN_DYNAMIC_SHAPE_MATRICES "Default" CACHE STRING "Use dynamically allocated shape matrices")
 set_property(CACHE OGS_EIGEN_DYNAMIC_SHAPE_MATRICES
              PROPERTY STRINGS "Default" "ON" "OFF")
@@ -192,6 +199,11 @@ endif()
 if (EIGEN_NO_DEBUG)
     add_definitions(-DEIGEN_NO_DEBUG)
 endif()
+
+if (EIGEN_DONT_VECTORIZE)
+    add_definitions(-DEIGEN_DONT_VECTORIZE)
+endif()
+
 if(OGS_EIGEN_DYNAMIC_SHAPE_MATRICES_INTERNAL)
     add_definitions(-DOGS_EIGEN_DYNAMIC_SHAPE_MATRICES)
 endif()

--- a/ProcessLib/HydroMechanics/Tests.cmake
+++ b/ProcessLib/HydroMechanics/Tests.cmake
@@ -89,7 +89,7 @@ AddTest(
     WRAPPER time
     TESTER vtkdiff
     REQUIREMENTS NOT OGS_USE_MPI
-    ABSTOL 5e-14 RELTOL 1e-16
+    ABSTOL 1e-13 RELTOL 1e-16
     DIFF_DATA
     expected_square_1e2_UC_late_pcs_0_ts_10_t_1000.000000.vtu square_1e2_UC_late_pcs_0_ts_10_t_1000.000000.vtu displacement displacement
     expected_square_1e2_UC_late_pcs_0_ts_10_t_1000.000000.vtu square_1e2_UC_late_pcs_0_ts_10_t_1000.000000.vtu pressure pressure

--- a/Tests/NumLib/TestComponentNorms.cpp
+++ b/Tests/NumLib/TestComponentNorms.cpp
@@ -143,7 +143,7 @@ TEST(MPITest_NumLib, ComponentNormMultiComponent1)
     unsigned const num_components = 3;
     auto const norm_type = MathLib::VecNormType::NORM1;
     auto const tolerance =
-        num_components * 600 * std::numeric_limits<double>::epsilon();
+        num_components * 1700 * std::numeric_limits<double>::epsilon();
 
     do_test(num_components, norm_type, tolerance,
             [](double n_total, double n) { return n_total + n; },
@@ -159,7 +159,7 @@ TEST(MPITest_NumLib, ComponentNormMultiComponent2)
     unsigned const num_components = 3;
     auto const norm_type = MathLib::VecNormType::NORM2;
     auto const tolerance =
-        num_components * 20 * std::numeric_limits<double>::epsilon();
+        num_components * 22 * std::numeric_limits<double>::epsilon();
 
     do_test(num_components, norm_type, tolerance,
             [](double n_total, double n) { return n_total + n*n; },

